api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/ParametrizedLayer.kt

```diff
@@ -0,0 +1,39 @@
+package org.jetbrains.kotlinx.dl.api.core.layer
+
+import org.jetbrains.kotlinx.dl.api.core.shape.numElements
+import org.tensorflow.Session
+import org.tensorflow.op.core.Variable
+
+public interface ParametrizedLayer {
+    /**
+     * Layer's variables
+     */
+    public val variables: List<VariableDto>
```

Comment:
By having it on layer explicitly, `weight` setter/getter can be implemented as an extension function outside of the `Layer` to not clutter its code:
```
/**
 *  Layer's weights
 *  Require parent model to be set on layer
 */
public var Layer.weights: Map<String, Array<*>>
    get() {
        if (this !is ParametrizedLayer) return emptyMap()

        val model = parentModel
        requireNotNull(model) { "Layer '$name' is not related to any model" }

        val variablesOrder = variables
        val runner = model.session.runner()
        variablesOrder.map { it.variable }.forEach(runner::fetch)
        val weights = runner.run().map { it.convertTensorToMultiDimArray() }

        return variablesOrder.map { it.name }.zip(weights).toMap()
    }
    set(weights) {
        if (this !is ParametrizedLayer) return

        val model = parentModel
        requireNotNull(model) { "Layer '$name' is not related to any model" }

        for ((name, value) in weights) {
            model.assignVariable(name, value)
        }
    }
```

and removed from all of the layers (in most of the cases it is simple 
``` 
override var weights: Map<String, Array<*>>
        get() = emptyMap()
        set(value) = assignWeights(value)
```
), along with `Layer::extractWeights`/`Layer::assignWeights`.